lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -674,83 +843,446 @@ export default function getAPIImplementation(): APIInterface {
+        user with email e2.
+
+        We rejected this idea cause if auto account linking is switched off, then someone else can sign up 
+        with google using e2. This is OK as it would not link (since account linking is switched off). 
+        But, then if account linking is switched on, then the google sign in (and not sign up) with e2 
+        would actually cause it to be linked with the e1 account.
+    */
+
+    // we allow setup of unverified account info only if the session user has the same account info
+    // and it is verified
+
+    if (accountInfo.email !== undefined) {
+        let foundVerifiedEmail = false;
+        for (const lM of sessionUser?.loginMethods) {
+            if (lM.hasSameEmailAs(accountInfo.email) && lM.verified) {
+                foundVerifiedEmail = true;
+                break;
             }
+        }
 
-            let session = await Session.createNewSession(
-                options.req,
-                options.res,
-                tenantId,
-                emailPasswordRecipeUser.recipeUserId,
-                {},
-                {},
-                userContext
-            );
-            return {
-                status: "OK",
-                session,
-                user: response.user,
-            };
-        },
-    };
-}
+        if (!foundVerifiedEmail) {
+            throw new SignUpError({
+                status: "SIGN_UP_NOT_ALLOWED",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_010)",
+            });
+        }
+    }
+};
+
+const checkIfFactorUserBeingCreatedCanBeLinkedWithSessionUser = async (
+    tenantId: string,
+    sessionUser: User,
+    accountInfo: { email: string },
+    userContext: UserContext
+) => {
+    if (!sessionUser.isPrimaryUser) {
+        const canCreatePrimary = await AccountLinking.getInstance().recipeInterfaceImpl.canCreatePrimaryUser({
+            recipeUserId: sessionUser.loginMethods[0].recipeUserId,
+            userContext,
+        });
+
+        if (canCreatePrimary.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+            // Session user is linked to another primary user, which means the session is revoked as well
+            throw new SessionError({
+                type: SessionError.TRY_REFRESH_TOKEN,
+                message: "Session may be revoked",
+            });
+        }
+
+        if (canCreatePrimary.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR") {
+            throw new SignUpError({
+                status: "SIGN_UP_NOT_ALLOWED",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_011)",
+            });
+        }
+    }
+
+    // Check if the linking with session user going to fail and avoid user creation here
+    const users = await listUsersByAccountInfo(tenantId, accountInfo, true, userContext);
```

Comment:
call the recipe implmentation function like we do everywhere else this is called

Reply:
Done